### DISCUSSION TITLE
Modeling

### DISCUSSION THREAD

**Discussion by tommy921092** [sensedeep/dynamodb-onetable#54] [Q&A] [answered]
using typescript, how can I model an attribute as Object or Array of Object?
And, Is there a better way to construct an entity with adjacency list pattern?
for example a top level entity called user, and under the user entity there is userProfile, userPayment, etc

**Comment by mobsense** [comment, marked as answer]
For the schema, there is nothing specific for typescript. Just `type: Object`

```
        Account: {
            pk:             { type: String, value: 'account:${id}' },
            sk:             { type: String, value: 'account:' },
            activity:       { type: Date },
            cardDetails:           { type: Object },  
```

I can imagine it would be nice to have your own type and be able to use that in the schema and have the TypeScript declarations propagate that. I've not checked that out -- will let you know.

You do an adjacency list as either a nested object inside the User ... like cardDetails above.

Alternatively, you can create a different entity that shares the same PK and leading SK. Then you can retrieve using a `begins` with on the query.

```
User: {
    pk: {type: String, value: 'user:${id}'}
    sk: {type: String, value: 'user#'}
}
UserProfile: {
    pk: {type: String, value: 'user:${id}'}
    sk: {type: String, value: 'user#profile'
}

let items = await table.queryItems({pk: `user:${id}`}, {sk: {begins: 'user#'}}
```

**Reply by tommy921092** [reply]
Yes I would like the Object to be of some declaration.
e.g. 
`cardDetails: {type: Object, schema: {name: {type: String}, details: {type: String}}}`
Thank you very much.

**Comment by mobsense** [comment]
Okay, the issue is a bit more subtle and detailed than it seemed at first.

Here is some results of our investigation so far. All input very welcome.

We have 3 issues:

1. How to update a single field in a map property.
2. How to specify nested schema map properties.
3. How to express (2) via TypeScript types.

Regarding (1) -- how to update a single property in a map

Currently, an update will update the entire map

Given schema:

    profile: {  type: Object }

Then

    User.update({... profile: newProfile })

This will replace the entire profile map. We currently don't have a way to do:

    User.update({"profile.status": "active"})

We're looking into this now, as both use cases are valid: i.e. updating the entire map or just a single field. I'd like to see both syntaxes. However, this may require (2) below as we'd need to understand the schema for the nested property.

Regarding (2) -- How to express nested schema

Currently nested schemas are not supported. i.e you cannot do the following as you suggest (yet):

    profile: {  type: Object, schema: { status: { type String } } }

Currently maps (type Object), are schema-less maps. You can write anything and we do not validate or apply any of the schema field level magic to the nested properties. I'd like to see us do this.  If we did this, then you could use the new syntax in (1) to update a single field in a map AND have validations, value templates, maps, enumerations, etc applied. You could also then update the entire map. If we did (3) below for TypeScript, updating the entire map would the be a typed operation and missing a field would be a compilation time error (sweet).

Regarding (3) -- How to express with TypeScript. It may be possible to convert nested  schema into TypeScript types. There are many questions. Should these be nested local types or should the nested schema be mapped to top level TypeScript types. Unclear if the TypeScript dynamic type magic in Model.d.ts can do this or not.

Thoughts?
